### PR TITLE
Preventing stack overflow crash while reading a huge throwable file

### PR DESCRIPTION
**Pull request by MatkovIvan**
* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Preventing stack overflow crash while reading a huge throwable file.

## Related PRs or issues

#855

### CODE REVIEW

File: apps/sasquatch/src/main/res/values/crashes.xml [hunk 1]
```diff
@@ -6,6 +6,8 @@
     <string name="description_crash_divide_by_0" tools:ignore="MissingTranslation">Divide by 0</string>
     <string name="title_test_ui_crash" tools:ignore="MissingTranslation">Test some UI crash</string>
     <string name="description_test_ui_crash" tools:ignore="MissingTranslation">A UI coding mistake, has inner exception</string>
+    <string name="title_deep_cause_exception_crash" tools:ignore="MissingTranslation">A exception with plenty of causes</string>
+    <string name="description_deep_cause_exception_crash" tools:ignore="MissingTranslation">The exception that caused the exception that caused the exception that caused the exception…</string>
```
Comment (jwallra) [inline apps/sasquatch/src/main/res/values/crashes.xml]: 😂

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java [hunk 2]
```diff
@@ -752,10 +752,10 @@ ErrorReport buildErrorReport(ManagedErrorLog log) {
                     ErrorReport report = ErrorLogHelper.getErrorReportFromErrorLog(log, throwable);
                     mErrorReportCache.put(id, new ErrorLogReport(log, report));
                     return report;
-                } catch (ClassNotFoundException ignored) {
-                    AppCenterLog.error(LOG_TAG, "Cannot read throwable file " + file.getName(), ignored);
-                } catch (IOException ignored) {
-                    AppCenterLog.error(LOG_TAG, "Cannot access serialized throwable file " + file.getName(), ignored);
+                } catch (ClassNotFoundException | StackOverflowError e) {
```
Comment (jwallra) [inline sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java]: Should these exceptions really have the same message?  I think we can be more explicit here so we can investigate more precisely when for support issues.

File: apps/sasquatch/src/main/res/values/crashes.xml (hunk 1 above)
Comment (jaelim-ms) [reply apps/sasquatch/src/main/res/values/crashes.xml]: 😂

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java (hunk 2 above)
Comment (dhei) [reply sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java]: `e` is logged in `AppCenterLog.error(LOG_TAG, "Cannot read throwable file " + file.getName(), e)` so it should be fine to reuse the same message.

File: apps/sasquatch/src/main/res/values/crashes.xml (hunk 1 above)
Comment (dhei) [reply apps/sasquatch/src/main/res/values/crashes.xml]: naming suggestion: something like "deeply_nested_exception_crash"

**Comment by guperrot** [comment]
Also another comment but might separate PR, the issue caused empty json files to be created and they are never cleaned up, we should also make a patch to delete the JSON file when we fail to read it for any reason.

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java (hunk 2 above)
Comment (guperrot) [reply sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java]: I would say we can even merge the 2 catch remaining catch blocks as the difference between messages is not useful.

**Comment by MatkovIvan** [comment]
Well, now when SDK try to send it without throwable, it crashes with
```
java.lang.OutOfMemoryError: Failed to allocate a 150994952 byte allocation with 25165824 free bytes and 103MB until OOM
        at org.json.JSONStringer.newline(JSONStringer.java:360)
        at org.json.JSONStringer.beforeKey(JSONStringer.java:390)
        at org.json.JSONStringer.key(JSONStringer.java:374)
        at org.json.JSONObject.writeTo(JSONObject.java:723)
        at org.json.JSONStringer.value(JSONStringer.java:237)
        at org.json.JSONArray.writeTo(JSONArray.java:613)
        at org.json.JSONStringer.value(JSONStringer.java:233)
        at org.json.JSONObject.writeTo(JSONObject.java:723)
        at org.json.JSONStringer.value(JSONStringer.java:237)
        at org.json.JSONArray.writeTo(JSONArray.java:613)
        at org.json.JSONStringer.value(JSONStringer.java:233)
...
        at org.json.JSONObject.toString(JSONObject.java:716)
        at com.microsoft.appcenter.http.DefaultHttpClient.doHttpCall(DefaultHttpClient.java:222)
        at com.microsoft.appcenter.http.DefaultHttpClient.doCall(DefaultHttpClient.java:152)
```

**Pull request merged**